go-controller/pkg/util/external_gw_conntrack.go

```diff
@@ -0,0 +1,318 @@
+//go:build linux
+// +build linux
+
+package util
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	"github.com/mdlayher/ndp"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+
+	kapi "k8s.io/api/core/v1"
+	utilapierrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+)
+
+const msgTries = 10
+
+// inspired by arping timeout
+var msgTimeout = 500 * time.Millisecond
+
+func findInterfaceForDstIP(dstIP string) (*net.Interface, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	netIP := net.ParseIP(dstIP)
+	if netIP == nil {
+		return nil, fmt.Errorf("failed to parse ip: %w", err)
+	}
+
+	isDown := func(iface net.Interface) bool {
+		return iface.Flags&net.FlagUp == 0
+	}
+	hasAddressInNetwork := func(iface net.Interface) bool {
+		addrs, err := iface.Addrs()
+		if err != nil {
+			return false
+		}
+		for _, a := range addrs {
+			if ipnet, ok := a.(*net.IPNet); ok {
+				if ipnet.Contains(netIP) {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	for _, iface := range ifaces {
+		if isDown(iface) {
+			continue
+		}
+		if !hasAddressInNetwork(iface) {
+			continue
+		}
+		return &iface, nil
+	}
+	return nil, errors.New("no usable interface found")
+}
+
+func readNDPMsg(msg ndp.Message) (targetIP netip.Addr, mac net.HardwareAddr, err error) {
+	// Expect a neighbor advertisement message with a target link-layer
+	// address option.
+	na, ok := msg.(*ndp.NeighborAdvertisement)
+	if !ok {
+		err = fmt.Errorf("message is not a neighbor advertisement: %T", msg)
+		return
+	}
+	if len(na.Options) != 1 {
+		err = fmt.Errorf("expected one option in neighbor advertisement")
+		return
+	}
+	lla, ok := na.Options[0].(*ndp.LinkLayerAddress)
+	if !ok {
+		err = fmt.Errorf("option is not a link-layer address: %T", msg)
+		return
+	}
+	// target ip doesn't have a zone set, return ip without a zone to compare
+	return na.TargetAddress.WithZone(""), lla.Addr, nil
+}
+
+// getIPv6MacOnIface tries to resolve as many ips as possible.
+// Errors that prevent only 1 ip from being resolved are logged and not returned.
+// When an error is returned, some MACs may still be resolved and returned too.
+func getIPv6MacOnIface(info *ifaceWithTargetIPs) ([]net.HardwareAddr, error) {
+	// Set up an *ndp.Conn, bound to this interface's link-local IPv6 address.
+	c, _, err := ndp.Listen(info.iface, ndp.LinkLocal)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial NDP connection: %v", err)
+	}
+	defer c.Close()
```

Comment:
it looks like c.Close can return error. Should we log this error at least to just signal to us something is wrong with our NDP connection handler?

Reply:
always a good idea!